models/items.js

```diff
@@ -14,24 +14,51 @@ const addTagsToItem = async (itemData) => {
   try {
     const { itemId, itemType, tagPayload } = itemData;
     const batch = firestore.batch();
+    const hasItemTagsPromises = [];
+    const isNewTag = {};
     for (const tag of tagPayload) {
-      const tagData = await tagModel.doc(tag.tagId).get();
-      const levelData = await levelModel.doc(tag.levelId).get();
-      const itemTag = {
-        itemId,
-        itemType: itemType.toUpperCase(),
-        tagId: tag.tagId,
-        tagName: tagData.data().name,
-        tagType: tagData.data().type,
-        levelId: tag.levelId,
-        levelName: levelData.data().name,
-        levelValue: levelData.data().value,
-      };
-      const docid = itemTagsModel.doc();
-      batch.set(docid, itemTag);
+      hasItemTagsPromises.push(
+        itemTagsModel
+          .where("itemId", "==", itemId)
+          .where("tagId", "==", tag.tagId)
+          .where("levelId", "==", tag.levelId)
+          .get()
+          .then((doc) => {
+            isNewTag[tag.tagId] = doc.empty;
+          })
+      );
     }
+    /**
+     * In the above code, we are looping across every tag that is to be set and creating an Object that will
+     * finally be used to check if atleast ONE of the tags is a new tag. Because if every tag is an existing
+     * tag, it is unnecessary to proceed any further. The `isNewTag` Object will look like
+     * {
+     *    some_tag_id_1: false,
+     *    some_tag_id_2: true
+     * }
+     * `some_tag_id_2:true` represent a new tag and false in the vice-versa.
+     */
+
+    await Promise.all(hasItemTagsPromises);
+    const wasSuccess = Object.values(isNewTag).some((value) => value === true); // Atleast one new tag
+
+    if (!wasSuccess) return { itemId, wasSuccess }; // Returns if none of the tags are new
+
+    tagPayload.forEach((tag) => {
```

Comment:
The isNewTag object is not being updated, so the if statement will always be true. This means that all of the tags will be added to the database, even if they are not new. To fix this, you need to update the isNewTag object after each iteration of the loop

Reply:
the `isNewTag` Object has a structure like
```
{
    some_tag_id_1: false,
    some_tag_id_2: true,
    .
    .
    .
}
```
Before adding a new document, it checks if the value of the tagId is true or not, then and only then it adds. 

Maybe I am missing your point😅